### PR TITLE
feat: allow preorders in Disjoint

### PR DESCRIPTION
**Pull request by kbuzzard**
It's convenient to have a way of saying "coprime" in a unique factorization monoid; we can't currently use Disjoint because the divisibility preorder on a UFM isn't in general a partial order. 

[Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Coprime.20if.20product.20is.20squarefree/near/419341762)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alreadydone** [comment]
I don't think you want to put the divisibility preorder on a unique factorization monoid, as it conflicts with the existing order on ZZ, for example. I think the usual way is to talk about disjointness of [UniqueFactorizationMonoid.normalizedFactors](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/UniqueFactorizationDomain.html#UniqueFactorizationMonoid.normalizedFactors) instead. (Strangely though, mathlib doesn't have that [Multiset.Disjoint](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Multiset/Basic.html#Multiset.Disjoint) is equivalent to Disjoint for Multisets.) But it may also be useful to define a coprime predicate that applies to rings without GCDs (just like Disjoint doesn't require Inf).

**Comment by kbuzzard** [comment]
The whole point is not to assume all these weird extra things like `NormalisationMonoid`. There's a perfectly good notion of being disjoint/coprime in a general UFM.

**Comment by alreadydone** [comment]
I see, normalizedFactors is not good enough. I think we should use Associates instead, which already has a [partial order instance](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Associated.html#Associates.instPartialOrder).

**Comment by urkud** [comment]
OTOH, why not generalize the definition?

**Comment by urkud** [comment]
!bench

**Review by alreadydone — APPROVED** [review]
maintainer merge

**Comment by eric-wieser** [comment]
Just to check: is this definition the right notion of disjoint without transitivity? The old definition was `x ∩ ⊓ y ≤ ⊥`, which I think would mean something different.

If it is correct, can you update the docstring comment to mention that the strange definition also websites validity in preorders?

Finally; should we make the same change to Codisjoint?

File: Mathlib/Order/Disjoint.lean [hunk 1]
```diff
@@ -80,11 +80,16 @@ theorem Disjoint.mono_right : b ≤ c → Disjoint a c → Disjoint a b :=
...
+
+section PartialOrderBot
+
+variable [PartialOrder α] [OrderBot α] {a b c d : α}
+
 @[simp]
 theorem disjoint_self : Disjoint a a ↔ a = ⊥ :=
   ⟨fun hd ↦ bot_unique <| hd le_rfl le_rfl, fun h _ ha _ ↦ ha.trans_eq h⟩
 #align disjoint_self disjoint_self
-
```
Comment (eric-wieser) [inline Mathlib/Order/Disjoint.lean:87]: Please add back this line :)

File: Mathlib/Order/Disjoint.lean [hunk 2]
```diff
@@ -29,9 +29,9 @@ variable {α : Type*}
 
 section Disjoint
 
-section PartialOrderBot
+section PreOrderBot
```
Comment (YaelDillies) [inline Mathlib/Order/Disjoint.lean:32]:
```suggestion
section PreorderBot
```

**Comment by urkud** [comment]
> If it is correct, can you update the docstring comment to mention that the strange definition also ensures validity in preorders?

AFAIR, one of the reasons to change the definition was to avoid dependency on `SemilatticeInf` (which requires, e.g., `DecidableEq` for `Finset`s). I think that we should change the definition to `∀ x, x ≤ a → x ≤ b → IsBot x` so that it works for orders without bottom element, but this can be done later.

**Comment by eric-wieser** [comment]
> AFAIR, one of the reasons to change the definition was to avoid dependency on `SemilatticeInf` (which requires, e.g., `DecidableEq` for `Finset`s).

Yes, I was the one who made this change originally, and that was exactly my motivation at the time! What I'm suggesting is that we seemed to have acquired an extra motivation, so we should document it in the same place.

**Comment by kbuzzard** [comment]
My motivation for this has disappeared because, as has been pointed out, I can't use it for UniqueFactorizationMonoids anyway, as the divisibility preorder is not an instance and furthermore they may have another order anyway (the obvious examples being Nat and Int). So I'm happy to let this one drop. For UniqueFactorizationMonoid one can move to Associates if necessary.

**Pull request closed without merging**